File: app/providers/schema.py
"""Pydantic v2 models describing the YAML provider definition.

Minimal shape:
    name: stripe
    base_url: https://api.stripe.com
    auth:
      type: bearer_static
      env: STRIPE_SECRET_KEY

Full shape adds: description, docs_url, enabled, default_headers,
strip_response_headers, and a forwarding block with mode, allow/deny
rules, rate_limits, and transparent-mode tunables.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator


class WindowSchema(BaseModel):
    capacity: int = Field(gt=0)
    window_ms: int = Field(gt=0)


class AllowRuleSchema(BaseModel):
    method: str
    pattern: str
    category: str = "default"
    cache_ttl: int = 0
    sensitive: bool = False

    @field_validator("method")
    @classmethod
    def _upper(cls, v: str) -> str:
        return v.upper()


class ForwardingSchema(BaseModel):
    mode: Literal["transparent", "allowlist"] = "transparent"
    allow: list[AllowRuleSchema] = Field(default_factory=list)
    deny: list[str] = Field(default_factory=list)
    treat_writes_as_sensitive: bool = True
    default_cache_ttl: int = 0
    default_category: str = "default"
    rate_limits: dict[str, list[WindowSchema]] = Field(
        default_factory=lambda: {
            "default": [
                WindowSchema(capacity=50, window_ms=1000),
                WindowSchema(capacity=500, window_ms=60_000),
            ]
        }
    )


class AuthSchema(BaseModel):
    type: str
    model_config = {"extra": "allow"}


class ProviderSchema(BaseModel):
    name: str
    base_url: str
    description: str = ""
    docs_url: str = ""
    enabled: bool = True
    default_headers: dict[str, str] = Field(default_factory=dict)
    strip_response_headers: list[str] = Field(default_factory=list)
    auth: AuthSchema
    forwarding: ForwardingSchema = Field(default_factory=ForwardingSchema)

    @field_validator("name")
    @classmethod
    def _name_safe(cls, v: str) -> str:
        if not v or not all(c.isalnum() or c in "-_" for c in v):
            raise ValueError("name must be non-empty alphanumeric, dash, or underscore")
        return v.lower()

File: app/ts/src/providers/schema.ts
/**
 * Zod schemas mirroring app/providers/schema.py 1:1.
 *
 * Minimal shape:
 *   name: stripe
 *   base_url: https://api.stripe.com
 *   auth:
 *     type: bearer_static
 *     env: STRIPE_SECRET_KEY
 */

import { z } from "zod";

export const WindowSchema = z.object({
  capacity: z.number().int().positive(),
  window_ms: z.number().int().positive(),
});

export const AllowRuleSchema = z.object({
  method: z.string().transform((s) => s.toUpperCase()),
  pattern: z.string(),
  category: z.string().default("default"),
  cache_ttl: z.number().int().nonnegative().default(0),
  sensitive: z.boolean().default(false),
});

export const ForwardingSchema = z.object({
  mode: z.enum(["transparent", "allowlist"]).default("transparent"),
  allow: z.array(AllowRuleSchema).default([]),
  deny: z.array(z.string()).default([]),
  treat_writes_as_sensitive: z.boolean().default(true),
  default_cache_ttl: z.number().int().nonnegative().default(0),
  default_category: z.string().default("default"),
  rate_limits: z.record(z.array(WindowSchema)).default({
    default: [
      { capacity: 50, window_ms: 1000 },
      { capacity: 500, window_ms: 60_000 },
    ],
  }),
});

export const AuthSchema = z.object({ type: z.string() }).passthrough();

export const ProviderSchema = z.object({
  name: z
    .string()
    .min(1)
    .regex(/^[a-zA-Z0-9_-]+$/, "alphanumeric, dash, underscore only")
    .transform((s) => s.toLowerCase()),
  base_url: z.string().url(),
  description: z.string().default(""),
  docs_url: z.string().default(""),
  enabled: z.boolean().default(true),
  default_headers: z.record(z.string()).default({}),
  strip_response_headers: z.array(z.string()).default([]),
  auth: AuthSchema,
  forwarding: ForwardingSchema.default({}),
});

export type ProviderDef = z.infer<typeof ProviderSchema>;
export type AllowRule = z.infer<typeof AllowRuleSchema>;
export type ForwardingDef = z.infer<typeof ForwardingSchema>;
export type WindowDef = z.infer<typeof WindowSchema>;
